logging the metrics of reqeust/error counts and latency when request extension apiserver

rayowang opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
logging the metrics of reqeust/error counts and latency when request extension apiserver

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #117167 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-apiserver adds two new metrics `extension_apiserver_attempts_total` and `extension_apiserver_duration_seconds` that allow users to monitor requests to extension apiserver, split by result.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

rayowang committed:
add the metrics of reqeust/error counts and latency during the apiserver aggregation

alexzielenski commented:
/sig instrumentation
/cc @logicalhan
/triage accepted

rayowang commented:
/ok-to-test

logicalhan commented:
/triage accepted
/assign

logicalhan commented:
/cc @dgrisonnet

shyamjvs commented:
/ok-to-test
/priority important-soon

shyamjvs commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/metrics.go:
```diff
@@ -46,7 +50,36 @@ var x509InsecureSHA1Counter = metrics.NewCounter(
+			Help:           "Counter of extension apiserver attempts broken down by result. It can be either 'OK', 'Not Found', 'Service Unavailable' or 'Internal Server Error'.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"result"},
```
Just like webhooks, we should also add a dimension here on the name of the extension apiserver imo.

@liggitt @deads2k  - looking for your inputs from apimachinery side :)

logicalhan replied:
I wouldn't call it `attempts`, I prefer `requests`

logicalhan replied:
I meant `extension_apiserver_requests_total`.

sttts replied:
+1. The lack of apiservice name or something like that surprises.

shyamjvs commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:
```diff
@@ -105,6 +107,12 @@ func proxyError(w http.ResponseWriter, req *http.Request, error string, code int
+	extApiserverStart := time.Now()
+	ctx := req.Context()
+	defer func() {
+		recordExtApiserverMetrics(ctx, httpStatus, extApiserverStart)
```
nit: better to use the full name to be clear `recordExtensionApiserverMetrics`

shyamjvs commented:
/retest

mengjiao-liu commented:
/remove-area cloudprovider
/remove-area kubectl
/remove-area kubelet
/remove-area dependency
/remove-area provider/gcp
/remove-area release-eng

/remove-sig scheduling
/remove-wg structured-logging
/remove-sig cli
/remove-sig apps
/remove-sig architecture
/remove-sig auth
/remove-sig network
/remove-sig release
/remove-sig windows
/remove-sig cluster-lifecycle

mengjiao-liu commented:
/remove-area code-generation
/remove-sig cloud-provider
/remove-sig storage
/remove-sig node
/remove-sig autoscaling
/remove-area e2e-test-framework

mengjiao-liu commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go:
```diff
@@ -1106,3 +1110,69 @@ func backendKey() []byte { return readTestFile("server-key.pem") }
+			defer extApiserverAttemptsCounter.Reset()
+			aggregationStart := time.Now()
+			ctx := context.TODO()
+			time.Sleep(time.Duration(rand.Intn(100)) * time.Millisecond)
```
```suggestion
			time.Sleep(time.Duration(random.Intn(100)) * time.Millisecond)
```

logicalhan commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/metrics.go:
```diff
@@ -46,7 +50,36 @@ var x509InsecureSHA1Counter = metrics.NewCounter(
+
+	extApiserverLatency = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Name:           "extension_apiserver_duration_seconds",
```
`request_duration_seconds`

rayowang replied:
Haha, in fact, I also have this habit, but it was written by others with reference to webhook before, and now I have changed it, thanks.

logicalhan commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/metrics.go:
```diff
@@ -46,7 +50,36 @@ var x509InsecureSHA1Counter = metrics.NewCounter(
+
+	extensionApiserverLatency = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Name:           "request_duration_seconds",
```
same here: `extension_apiserver_request_duration_seconds`.

logicalhan reviewed (commented):
/lgtm
/approve

logicalhan commented:
Can you squash?

mengjiao-liu commented:
> Can you squash?

You can refer to this [page](https://www.kubernetes.dev/docs/guide/github-workflow/#squash-commits) to squash your commits. @rayowang

logicalhan commented:
/lgtm

mengjiao-liu commented:
> > > Can you squash?
> > 
> > 
> > You can refer to this [page](https://www.kubernetes.dev/docs/guide/github-workflow/#squash-commits) to squash your commits. @rayowang
> 
> ok, thanks, I'm done. I thought it would squash automatically when github merged, which is the case in other projects.

kubernetes repo doesn't squash automatically unless you type the `/label tide/merge-method-squash`.

/lgtm

shyamjvs commented:
@wojtek-t can you help approve this when you get a sec?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/metrics.go:
```diff
@@ -46,7 +50,36 @@ var x509InsecureSHA1Counter = metrics.NewCounter(
+
+	extensionApiserverLatency = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Name:           "extension_apiserver_request_duration_seconds",
```
we have `request_duration_seconds` which has a group label. Isn't that enough?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/metrics.go:
```diff
@@ -46,7 +50,36 @@ var x509InsecureSHA1Counter = metrics.NewCounter(
+		&metrics.HistogramOpts{
+			Name:           "extension_apiserver_request_duration_seconds",
+			Help:           "extension apiserver request duration in seconds broken out by result.",
+			Buckets:        metrics.ExponentialBuckets(0.001, 2, 15),
```
what's the ratio behind these pretty coarse buckets?

sttts replied:
For reference, request_duration_seconds has `Buckets: []float64{0.005, 0.025, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.25, 1.5, 2, 3, 4, 5, 6, 8, 10, 15, 20, 30, 45, 60}`.

sttts replied:
Same here: the lack of apiservice name or something like that surprises.

rayowang committed:
fix

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/metrics.go:
```diff
@@ -64,7 +64,7 @@ var (
 			Name:           "extension_apiserver_request_duration_seconds",
 			Help:           "extension apiserver request duration in seconds broken out by result.",
-			Buckets:        metrics.ExponentialBuckets(0.001, 2, 15),
+			Buckets:        []float64{0.005, 0.025, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.25, 1.5, 2, 3, 4, 5, 6, 8, 10, 15, 20, 30, 45, 60},
```
@benluddy @logicalhan ptal.

dgrisonnet replied:
I don't think we need that much precision for the extension apiserver. The following buckets could be enough to start with and we can adjust then later on if needed.
```
[]float64{0.005, 0.025, 0.05, 0.1, 0.2, 0.5, 1, 2, 5, 10, 25, 60}
```

benluddy replied:
Would it be more useful operationally to spend these timeseries on being able to distinguish between backends? With one or more remote aggregated apiservers, plus an apiextensions-apiserver, what shape will the distribution of request latencies have?

dgrisonnet replied:
Oh. When I first looked at the PR, I assumed it was distinguishing between backends via a `resource` label, but that's not the case. I agree with you that the metric is not very useful if we don't have that.

benluddy replied:
> distinguishing between backends via a resource label

If that's not practical here, we should at least be able to distinguish by group/version since that is available to the APIService controller when it instantiates a handlerProxy.

benluddy commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/metrics.go:
```diff
@@ -46,7 +50,36 @@ var x509InsecureSHA1Counter = metrics.NewCounter(
+
+func recordExtensionApiserverMetrics(ctx context.Context, httpStatus int, extensionApiserverStart time.Time) {
+	extensionApiserverFinish := time.Now()
+	extensionApiserverRequestCounter.WithContext(ctx).WithLabelValues(http.StatusText(httpStatus)).Inc()
```
The label value is the textual status name (via `http.StatusText`)? I think the numeric code (as label "code") would be preferable, and consistent with `apiserver_request_total`.

benluddy commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:
```diff
@@ -114,6 +120,7 @@ func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+			httpStatus = http.StatusNotFound
 			return
 		}
 		r.localDelegate.ServeHTTP(w, req)
```
Where is the HTTP status written by the CR handler captured? It looks like, as currently implemented, the metric will report 200 regardless of the status written by the delegate handler.

benluddy replied:
The same comment applies to the case where the response status is set by the upgrade aware proxy handler.

rayowang replied:
@benluddy The latest modification reads the status of http.ResponseWriter to record the status of the delegate handler. Please review again if you have time.

benluddy commented:
These metric names imply (to me) that they count only requests to an apiextensions-apiserver. Requests proxied to remote aggregated apiservers are being counted here, too. Do we intend to cover both in the same metric? I imagine that there will be a bimodal latency distribution based on whether or not a request needs to make an extra network hop.

dgrisonnet commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/metrics.go:
```diff
@@ -46,7 +50,36 @@ var x509InsecureSHA1Counter = metrics.NewCounter(
+var (
+	extensionApiserverRequestCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Name:           "extension_apiserver_requests_total",
```
Could you extract `extension_apiserver` from the name and put it in the `Subsytem` field? Similar to what is done in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L63. This makes sure that withing this file all the metrics will share the same prefix.

rayowang committed:
extract extension_apiserver from the name and put it in the Subsytem

rayowang committed:
use http code to consistent with apiserver_request_total

rayowang committed:
add a resource dimension to the metric in order to select aggregated apiservers

dgrisonnet commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:
```diff
@@ -105,15 +105,24 @@ func proxyError(w http.ResponseWriter, req *http.Request, error string, code int
+	extensionApiserverStart := time.Now()
+	ctx := req.Context()
+	defer func() {
+		recordExtensionApiserverMetrics(ctx, httpStatus, extensionApiserverStart, serviceName)
```
Passing the apiservice as the resource name will not make correlation between apiserver_requests_duration and extension_requests_duration possible since the values of the resource labels will be different.

@benluddy had a pretty neat suggestion in https://github.com/kubernetes/kubernetes/pull/118636#discussion_r1251215059, to use group/version instead

rayowang replied:
@dgrisonnet Group/version cannot be read in the code context, so in the latest modification I use serviceNamespace and serviceName to form the resource name. If you have time, please help to see if this is feasible.

shyamjvs commented:
@rayowang are you planning to get this change in soonish?

rayowang commented:
> are you planning to get this change in soonish?

This week I will continue to solve this problem.

shyamjvs commented:
@rayowang seems like we're close to the finish line modulo few comments - do you wanna address those soonish?

shyamjvs commented:
trying one last time - @rayowang are you still available to get this PR to the finish line?

rayowang committed:
1.record HTTP status written by the CR handler captured
2.replace serviceName with "serviceNamespace/serviceName"
3.fix conflicts dute to rebase upstream/master

rayowang commented:
> trying one last time - @rayowang are you still available to get this PR to the finish line?

Sorry, there was a problem with my github account before, so it took some time. Now I have made changes according to the review suggestions. If you have time, please help me review it.

neolit123 commented:
/remove-sig cluster-lifecycle
/remove-area kubeadm
